### ISSUE TITLE
[BUG] Sync-only Authentication Callbacks not Working on Async Operations

### ISSUE THREAD

**Issue by Xdynix** [vitalik/django-ninja#1251]
**Describe the bug**

If an authentication callback can only works in sync context, then it will not work on async views.

Example Code:

```python
from ninja.security import django_auth

api = NinjaAPI()


@api.get("/foobar", auth=django_auth)
async def foobar() -> str:
    return "foobar"
```

Accessing the view will raise: `django.core.exceptions.SynchronousOnlyOperation: You cannot call this from an async context - use a thread or sync_to_async.`

This is because accessing `request.user` involves DB query (when using Django's default session engine) and can only be run in sync context. But in `AsyncOperation._run_authentication()` it didn't switch the context for the authentication callback.

**Versions (please complete the following information):**
 - Python version: [e.g. 3.6]: 3.12
 - Django version: [e.g. 4.0]: 5.0.7
 - Django-Ninja version: [e.g. 0.16.2]: 1.2.2
 - Pydantic version: [e.g. 1.9.0]: 2.8.2

Note you can quickly get this by runninng in `./manage.py shell` this line:
```
import django; import pydantic; import ninja; django.__version__; ninja.__version__; pydantic.__version__
```

**Comment by Xdynix** [comment]
This issue is probably beyond the scope of Django Ninja, but I think it would be helpful if the documentation mentioned it.

My current workaround is to create a decorator like this:

```python
from asyncio import get_running_loop
from collections.abc import Callable
from concurrent.futures import ThreadPoolExecutor
from functools import wraps
from typing import Any, TypeVar, cast

F = TypeVar("F", bound=Callable[..., Any])


def ensure_sync_context(func: F) -> F:
    """Run function with its own thread when in async context."""

    @wraps(func)
    def wrapper(*args: Any, **kwargs: Any) -> Any:
        try:
            get_running_loop()
        except RuntimeError:
            return func(*args, **kwargs)
        else:
            with ThreadPoolExecutor(max_workers=1) as executor:
                return executor.submit(func, *args, **kwargs).result()

    return cast(F, wrapper)
```

Then for callbacks that I am sure will perform async-unsafe operations, I will add this decorator (such as `SessionAuth.authenticate()`).

**Comment by bellini666** [comment]
> This issue is probably beyond the scope of Django Ninja

I would argue that [this line](https://github.com/vitalik/django-ninja/blob/9e1e5f60e34375852b066759ff108c1ce76e769d/ninja/operation.py#L382) could be changed from:

```python
result = callback(request)
```

to:

```python
result = await sync_to_async(callback)(request)
```

Would be an actual safe way to solve this?

I mean, an authentication code will usually retrieve something from the database. If we are inside `AsyncOperation`, we are running asynchronously, so we should be using `sync_to_async` to call this kind of callback

NOTE: The sync operation already does [async_to_sync](https://github.com/vitalik/django-ninja/blob/9e1e5f60e34375852b066759ff108c1ce76e769d/ninja/operation.py#L210) in this line

**Comment by Xdynix** [comment]
> ... changed from:
> ```python
> result = callback(request)
> ```
> to:
> ```python
> result = await sync_to_async(callback)(request)
> ```

This does work. However, one argument I can think of is that it forces the callback to do a context switch, resulting in some performance loss. If we keep the status quo, we can give the callback itself the choice of whether to switch or not.

**Comment by bellini666** [comment]
> This does work. However, one argument I can think of is that it forces the callback to do a context switch, resulting in some performance loss. If we keep the status quo, we can give the callback itself the choice of whether to switch or not.

Yes, but the performance loss in this case is negligible.

Also, as I commented earlier, "authentication" usually involves doing something I/O-bound (i.e. db query, request to another service, call to redis, etc), which would either block the mainloop or raise an error (`SynchronousOnlyOperation` for django, to prevent the blocking). Not doing an I/O-bound operation is most likely a corner case exception, IMHO.

Also, to prevent the context switch, an async callback could be used instead.

**Comment by Xdynix** [comment]
I totally agree with your point, I don't mind the context switch here either. I’m mostly raising the argument for completeness and to document the trade-offs.
